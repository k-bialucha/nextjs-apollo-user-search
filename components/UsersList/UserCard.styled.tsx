import styled from 'styled-components';
import { IAppTheme } from '../../styling/theme';

export const Container = styled.div`
  border-bottom: 2px solid
    ${(props: { theme: IAppTheme }) => props.theme.border};

  margin: 0.8em 0;
  padding: 0.5em;
`;

export const HighlightedText = styled.span`
  color: ${(props: { theme: IAppTheme }) => props.theme.main};
  font-weight: 600;
`;
