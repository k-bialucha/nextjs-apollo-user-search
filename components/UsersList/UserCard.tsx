import React from 'react';
import Link from 'next/link';

import { Container, HighlightedText, ProfileButton } from './UserCard.styled';

interface Props {
  name: string;
}

const UserCard: React.FC<Props> = ({ name }) => {
  return (
    <Container>
      Name: <HighlightedText>{name}</HighlightedText>
      <Link href={`/user/${name}`}>
        <ProfileButton type="button">show more</ProfileButton>
      </Link>
    </Container>
  );
};

export default UserCard;
