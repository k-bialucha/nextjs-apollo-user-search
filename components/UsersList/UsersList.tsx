import React from 'react';
import gql from 'graphql-tag';
import { useQuery } from '@apollo/react-hooks';

import UserCard from './UserCard';
import Card from '../Card';

import { HighlightedText } from './UserCard.styled';

export interface Query {
  type: 'login' | 'name' | 'email';
  key: string;
}

interface UserData {
  login: string;
  email: string;
}

export const SEARCH_USERS_QUERY = gql`
  query SearchUsers($query: String!, $first: Int!) {
    search(query: $query, type: USER, first: $first) {
      edges {
        node {
          ... on User {
            login
            email
          }
        }
      }
    }
  }
`;

interface Props {
  query: Query;
}

const UsersList: React.FC<Props> = ({ query }) => {
  const { loading, error, data } = useQuery(SEARCH_USERS_QUERY, {
    variables: {
      query: query.key,
      first: 10,
    },
    notifyOnNetworkStatusChange: true,
  });

  return (
    <Card title="Users">
      {!query.key && <p>Type a query in the box above to see some results.</p>}
      {query.key && (
        <p>
          Results for query: <HighlightedText>{query.key}</HighlightedText>.
        </p>
      )}
      {data?.search?.edges.map((edge: { node: UserData }) => (
        <UserCard key={edge.node.login} name={edge.node.login} />
      ))}
    </Card>
  );
};

export default UsersList;
