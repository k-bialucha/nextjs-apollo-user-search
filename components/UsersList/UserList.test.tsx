import React from 'react';
import { shallow } from 'enzyme';

import * as ApolloHooks from '@apollo/react-hooks';

import UsersList from './UsersList';

import UserCard from './UserCard';

jest.spyOn(ApolloHooks, 'useQuery').mockImplementation(jest.fn());
const useQueryMock = ApolloHooks.useQuery as jest.Mock;

const fakeData = {
  search: {
    edges: [
      {
        node: {
          login: 'adam',
          email: 'adam.french@cpanel.net',
          __typename: 'User',
        },
        __typename: 'SearchResultItemEdge',
      },
      {
        node: { login: 'adamwathan', email: '', __typename: 'User' },
        __typename: 'SearchResultItemEdge',
      },
    ],
    __typename: 'SearchResultItemConnection',
  },
};

describe('UserList', () => {
  it('renders a list of UserCard items', () => {
    useQueryMock.mockReturnValue({ data: fakeData });

    const wrapper = shallow(
      <UsersList query={{ key: 'fake-query', type: 'login' }} />
    );

    const userItems = wrapper.find(UserCard);

    expect(userItems.length).toBe(fakeData.search.edges.length);
  });
});
