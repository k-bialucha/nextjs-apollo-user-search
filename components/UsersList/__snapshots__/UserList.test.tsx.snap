// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList matches the snapshot in empty query state 1`] = `
<Card
  title="Users"
>
  <p>
    Type a query in the box above to see some results.
  </p>
</Card>
`;

exports[`UserList matches the snapshot in normal state 1`] = `
<Card
  title="Users"
>
  <p>
    Results for query: 
    <styled.span>
      fake-query
    </styled.span>
    .
  </p>
  <UserCard
    key="some-login"
    name="some-login"
  />
  <UserCard
    key="fancy-login"
    name="fancy-login"
  />
</Card>
`;
